### PR TITLE
Xdai deployment

### PR DESCRIPTION
**Pull request by fleupold**
This PR upates the dependencies to contain xDAI ready contracts and performs an xDAI migration.

```
yarn truffle migrate --network xdai
```

Note that the addresses for the Safe contracts (GnosisSafeProxyFactory, GnosisSafe, MultiSend) have been added to networks.json by hand, partially because of https://github.com/gnosis/safe-contracts/issues/231 and partially due to https://github.com/gnosis/dex-liquidity-provision/issues/517

### Test Plan
```
yarn truffle exec scripts/complete_liquidity_provision.js --baseTokenId=1 --quoteTokenId=8 --lowestLimit=300 --highestLimit=400 --currentPrice=340 --masterSafe=$MASTER_SAFE --depositBaseToken=0 --depositQuoteToken=0 --numBrackets=10 --network=$NETWORK_NAME
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,11 +43,11 @@
     "truffle-plugin-verify": "^0.4.0"
   },
   "dependencies": {
-    "@gnosis.pm/dex-contracts": "^0.4.2",
+    "@gnosis.pm/dex-contracts": "^0.4.3",
     "@gnosis.pm/owl-token": "^3.1.0",
-    "@gnosis.pm/safe-contracts": "v1.1.1-dev.2",
+    "@gnosis.pm/safe-contracts": "v1.2.0",
```
Comment (fedgiac) [inline package.json:48]: I already did the research and I believe that this version switch is safe in this case.
On the other end, I think the change deserves its own PR, and without any check it's very risky because the code in this repo would use the artifacts of a Safe v1.2.0 to manage the currently existing Safe v1.1.1, and there are no unit tests covering this scenario.

File: scripts/utils/gnosis_safe_server_interactions.js [hunk 2]
```diff
@@ -13,6 +13,7 @@ module.exports = function (web3 = web3, artifacts = artifacts) {
   const linkPrefix = {
     rinkeby: "rinkeby.",
     mainnet: "",
+    xdai: "xdai",
```
Comment (fedgiac) [inline scripts/utils/gnosis_safe_server_interactions.js:16]:
```
$ curl  https://xdai.gnosis-safe.io/app/#/welcome
curl: (6) Could not resolve host: xdai.gnosis-safe.io
```
There is currently no frontend at this address, is this expected to change?

**Review by fedgiac — COMMENTED** [review]
Thanks for putting in the work.

I can't see the [contract code](https://blockscout.com/poa/xdai/address/0x473E1D34458F26BFeB40f304a13c80FA291d0Abc/contracts) for the fleet factories on xDAI, can you include the xDAI contract verification procedure?

The other contracts look good (but see comments).

File: scripts/utils/gnosis_safe_server_interactions.js (hunk 2 above)
Comment (fedgiac) [reply scripts/utils/gnosis_safe_server_interactions.js:16]: It's good however that the [transaction service exists](https://safe-transaction.xdai.gnosis.io/api/v1/about/) .

File: networks.json [hunk 3]
```diff
@@ -1,4 +1,30 @@
...
+    },
+    "100": {
+      "links": {
+        "IdToAddressBiMap": "0x048E53A455a058462eA58442E1D94Fbc955495cB",
+        "IterableAppendOnlySet": "0x57E6B987c2ccd421859A244dd22a0D5a62D88f91"
+      },
+      "address": "0x25B06305CC4ec6AfCF3E7c0b673da1EF8ae26313",
+      "transactionHash": "0xd5e7da686ace1ae7f1917ad54b8610ee936bdc900e56b495739795fba3d5e7c1"
+    }
+  },
```
[the remark this answers is not in the document]
Comment (fleupold) [reply networks.json:27]: They got included when I ran `yarn networks-extract`to get the new xDAI addresses. I will remove them manually but it seems a bit weird that we have to manually adjust the networks.json in so many places.

File: scripts/utils/gnosis_safe_server_interactions.js (hunk 2 above)
Comment (fleupold) [reply scripts/utils/gnosis_safe_server_interactions.js:16]: Yes, the frontend will be deployed ETA next week

File: networks.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (fedgiac) [reply networks.json]: Yeah, I agree. 
A good solution might be to change `@gnosis.pm/util-contracts/src/inject_network_info.js` so that if you are trying to inject a network into an artifact that already has a (different) address then the execution fails. In this way, we can keep all addresses we use in this file, and if some artifacts get a different address with an update we get a failing CI. (Which is good, since changes to these addresses would most likely command some manual check.)

[the remark this answers is not in the document]
Comment (fleupold) [reply networks.json]: Usually, you run `yarn networks-inject` before migrating though, so the address would be set as there is non already. I think we should try to avoid keeping our own copy of the SafeArtefact in the repo and instead of using 

```
const GnosisSafe = artifacts.require("./GnosisSafe.sol")
```

use as Richard suggested

```
const GnosisSafeBuildInfo = require("@gnosis.pm/safe-contracts/build/contracts/GnosisSafe.json")
const GnosisSafe = truffleContract(GnosisSafeBuildInfo)
```

and somehow make sure development deployments are recorded in the dependent npm artefact.

**Comment by fleupold** [comment]
> can you include the xDAI contract verification procedure?

Same as on etherscan. Click verify/publish on the Contract address and add the code (can be taken from Etherscan as well as the compiler version). Verified contracts are here:

https://blockscout.com/poa/xdai/address/0xE91c3d95629465e597CC35ee407d0f11aE19646E/contracts
https://blockscout.com/poa/xdai/address/0x473E1D34458F26BFeB40f304a13c80FA291d0Abc/contracts

File: scripts/utils/gnosis_safe_server_interactions.js [hunk 4]
```diff
@@ -132,7 +133,7 @@ module.exports = function (web3 = web3, artifacts = artifacts) {
    *
    * @param {Address} multisigAddress Address of the multisig Safe for
    * which to retrieve the nonce.
-   * @param {string} network Either rinkeby or mainnet.
+   * @param {string} network Either rinkeby, dai or mainnet.
```
Comment (bh2smith) [inline scripts/utils/gnosis_safe_server_interactions.js:136]:
```suggestion
   * @param {string} network Either rinkeby, xdai or mainnet.
```

**Review by bh2smith — APPROVED** [review]
This looks fantastic! Nice work and no invasive changes! 🥇

**Review by fedgiac — APPROVED** [review]
